### ISSUE TITLE
What is the "fate" of Aurelia samples

### ISSUE THREAD

**Issue by adriatic** [auth0-community/auth0-aurelia-samples#22]
I tried to help with these samples a year ago, when just about everything going in Auth0 was in flux, including the basic services. I spent some time interacting with Ryan @chenkie but did not get much done, as we were both too busy and too many issues related to whether to use the Auth0 Lock hosted login were not obvious to either of us.

I just checked the **[Aurelia samples](https://github.com/auth0-community/auth0-aurelia-samples)** again and realized that the single [01-Login](https://github.com/auth0-community/auth0-aurelia-samples/tree/master/01-Login)** sample uses too old version of Aurelia specific implementation of webpack.

If there is the interest in making this one sample ___current___ and making the set of samples ___complete___ (like **[angular-samples](https://github.com/auth0-samples/auth0-angular-samples)**), I would be happy to help if someone sufficiently experienced with Auth0 can join me in that effort.

**Comment by smoore4moma** [comment]
I will see what I can do.  Just went live with a public Aurelia app (https://maid.moma.org), but I have a private  version of the same app at cems.moma.org that uses Auth0 to login.

**Comment by adriatic** [comment]
Hi @smoore4moma - now I why did you disappear from the AUCS project 😏. Thanks for your offer to help and let me explain a bit more about my **[proposal above](https://github.com/auth0-community/auth0-aurelia-samples/issues/22#issue-312105950)**:

The work in this context is envisioned as providing a "comfortable feeling" to Aurelia developers who may somehow find their way to this samples collection. So, by making the set of samples ___complete___, I meant to make it equivalent to this set of Angular samples

![image](https://user-images.githubusercontent.com/2712405/38443290-3a0d28f4-39b9-11e8-81ae-fde35fcb467f.png).

I am pretty well "armed" with Auth0 myself, so by asking for ___someone sufficiently experienced with Auth0___ I tried to ensure that I do not spend the rest of my time on some "odd corners" that might be hidden in less often use of Auth0 PaaS. 

Of course, if you are such guy, even better 😄

**Comment by aaguiarz** [comment]
Hi @adriatic 

I don't have much knowledge of Aurelia, but let me know how I can help on the Auth0 side.

Regards,

Andres

**Comment by adriatic** [comment]
Hello, @aaguiarz 

You seem to be the best suited collaborator then, as I need no help on Aurelia side and **will likely** need **some** help on the Auth0 side. 

Here are as few organizational issues, based on the plan to create all 6 samples (01 - 06) or even seven if we should create the '00' sample as well:

- Is it the standard practice to add the new code by issuing PRs? 
- Is there someone who responds to new PRs and merges them relatively fast
- the currently available sample **[01-login](https://github.com/auth0-community/auth0-aurelia-samples/tree/master/01-Login)** is a lot behind the current Aurelia framework. (for example, it uses ES6 / webpack 2.x build, where the current webpack is version 4 - dramatically different), Rather than **updating** this code, it would be a lot better to **replace** it.

I am engaged full time on various Aurelia tools ad libraries and I am big believer in Auth0 for many years. So, my intent here is to make the **[auth0-aurelia-samples](https://github.com/auth0-community/auth0-aurelia-samples)** as good as possible, both for future Aurelia developers visits as well for Auth0.

**Comment by adriatic** [comment]
In order to me most efficient (with respect to my schedule), I would think that the best approach would be to work on this project in "my" repository **[auth0-aurelia-samples](https://github.com/aurelia-auth0/auth0-aurelia-samples)** repository, where I would add you as a team member. This way you could clone this repo "wholesale" when finished.

**Comment by adriatic** [comment]
The last decision to be made is on building / tooling use. 

My recommendation would be to use **[aurelia-cli](https://github.com/aurelia/cli)** with webpack 4 bundler. This setup will make most Aurelia developers happy. 

Finally we need to make the decision on ES6 / Typescript based samples. Aurelia developers are split 50-50 here, so the optimal solution would be to make two parallel sets of samples.

**Comment by aaguiarz** [comment]
The best approach would be to have a PR per sample, so we can review/test/merge independently. I'll review it the code makes sense from an Auth0 usage perspective, and someone will help doing manual testing. Then we can merge.

I'm OK with you deleting most of the content in the 01-Login folder and start from scratch. The final should have the same UI as the other SPA Quickstarts, as we are adding UI automation to Quickstarts.

You can make the call regarding building/tooling. About ES6/TS, I think you should pick your preferred one, make that work, and after we are done we can evaluate if it make sense to add the other. I'd rather keep only one.

I think it would be better if you fork the current repo, work on it, and submit PRs to the Auth0 one.

We will also need to create documents that would go here https://auth0.com/docs/quickstart/spa/aurelia, is that some you would also want to do?

Thanks a lot for your help.

Andres

**Comment by adriatic** [comment]
Thanks for being quick and decisive - this is a great start. While I would be more comfortable working in my own repository, I fully understand (and will go along with) your preference. 

I spent all day yesterday trying to "replace aurelia parts" in the **[](https://github.com/auth0-community/auth0-aurelia-samples/tree/master/01-Login)** sample - and failed to make it work. Going the other way (adding Auth0 specific code) to a working Aurelia ES6/Webpack sample is an order of magnitude easier. While updating that sample, versus deleting it and adding the new one is really a git issue, I will take your preference here.

> We will also need to create documents that would go here https://auth0.com/docs/quickstart/spa/aurelia, is that some you would also want to do?

By all means, yes.

**Comment by aaguiarz** [comment]
Sounds great!

**Comment by adriatic** [comment]
In the case you wonder how did I suddenly appear here with the desire to help Auth0:

- I know both Matias and Eugenio from the days when Auth0 was very very small company, so my affection for Auth0 is well foundeed.
- My current project for Aurelia includes the work on Auth0 in Aurelia Community. In order to do this right, I want to first make a small contribution to Aurelia support in Auth0. Not only will I get to know more about Auth0, I most likely found a great guy (@aaguiarz in the case you wondered who would that guy be 😏 )

**Comment by adriatic** [comment]
@aaguiarz I would like to discuss a few auth0 specific issues - let me know if you have the time and desire to help me with that, please

**Comment by aaguiarz** [comment]
Yes, email me at andres.aguiar at auth0 dot com and we can schedule some time.

**Comment by adriatic** [comment]
Perfect, much appreciated. Let me just give you a brief peek into the subject I would like "master". The article **[what happens after the authentication request from a SPA application](https://github.com/longarone/global/issues/4#issue-315939720)** is a summary of my more than a year old "chat" with Joao Mangelo (Autho), focusing on the "flow of control" after my SPA app issues the authentication request to the auth0-js server. 

Today, I have a lot better definition of what I want use 

1. auth0.WebAuth (not the lock widget)
```
   this.auth0 = new auth0.WebAuth({
      clientID: AuthConfig.clientID,
      domain: AuthConfig.domain,
      responseType: 'token id_token',
      audience: AuthConfig.audience,
      redirectUri: AuthConfig.callbackURL
    });
```

**Comment by adriatic** [comment]
Thanks, Andrés for such quick response. You addressed only the question "how to preserve application state when using the auth0 server for authentication", answer to which is "use local storage". This could be quite bothersome of my app has to maintain a lot of such information. Before I can claim to "my users" that they should do authentication first, when there is no state information other than the app's "current location", I would appreciate if we can discuss it some more.

 A few comments:

1. The **[sample](https://github.com/auth0-samples/auth0-javascript-samples/tree/popup-mode-sample)** you just wrote is not the best guide to implement Login in the case of of Angular/Aurelia like applications that have strict  adherence to view / view model architectural pattern (for example the **[callback_popup.htm](https://github.com/auth0-samples/auth0-javascript-samples/blob/popup-mode-sample/01-Login/callback_popup.html)** mixes JavaScript and HTML code. 

2. A possibly silly question: why is the authorize request not issued in the form of a standard synchronous http request, like asking the auth0 server what is the time, or fetching some object that server can return. The answer to this question is the core of our discussion, so in order to minimize my impact on your time, I will come back here after getting full understanding of https://connect2id.com/learn/openid-connect document. Please let me know if you recommend some other source of information.

3. I believe that you see my intent - I want to make Auth0 Aurelia samples and tutorials better than any other collection, as I want to make the introduction of Auth0 to Aurelia community as rich as possible. My questions are partially induced by the observation that the average application developer knows very little about identity management.

**Summary** I would like to create my own answers to this discussion and run it by you before starting to do the real work on Aurelia samples, where I would like to present the conclusions from our discussion.

**Comment by aaguiarz** [comment]
>You addressed only the question "how to preserve application state when using the auth0 server for >authentication", answer to which is "use local storage"

In an SPA you need to handle state somehow. If the users refreshes the page, or deep link to the current anchor, they'd expect to see the same state. If you redirect to a different site and get back, you need to go to the same URL where you were, and that should handle loading the UI with the right state. The 'popup' approach lets you avoid the issue by not reloading the SPA page.

About 1), callback_popup.html, that is an utility html that is used to make Auth0 be able to send the securely the tokens to the parent frame once the window is closed. I understand it feels 'dirty' but I'd take it as part of the infrastructure and not part of the app.

About 2) because that's not why OIDC works, and the are a bunch of security reasons why it is that way. Tokens are always returned by redirecting to a whitelisted callback url. You can do that with a full redirect in the home page, or handle the redirect in an iframe like the popup example does.

**Comment by adriatic** [comment]
Everything you say, makes a lot of sense. I appreciate your willingness to discuss it with me and I hope you will stay "attached" to this Aurelia-Auth0 related project 😄

**Comment by adriatic** [comment]
## A very new question Andres

I am very keen on creating the new Aurelia based samples in the spirit of Aurelia - not just by replicating the current 01-Login code, which is very far from working with today's Aurelia (I can show to you all the reasons why not at a later time). 

At the moment please explain to the behavior, I am observing now:

After the successful authentication my app is "invoked" from the server with:
```
http://localhost:3000/#access_token=CSGmNaxb9MaahYC5HX0S3qcSgGtmOqxB&expires_in=7200&token_type=Bearer&state=dCn~HCfAVo6d8u1F3tFmawE96Q7dehCS&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IlJqWkJOVFkyTnpSQk5VVTVSakF5TkVZeFJEbEZSRUV4T0RWQlJFSTFOVUl4UmpOR1JEQkJPUSJ9.eyJuaWNrbmFtZSI6ImFjLWRlbW8iLCJuYW1lIjoiYWMtZGVtb0BhdXJlbGlhdG9vbHMuY29tIiwicGljdHVyZSI6Imh0dHBzOi8vcy5ncmF2YXRhci5jb20vYXZhdGFyL2E0ZWEyNThhYjk5ZWQ1NDQxNzFjOGNlMmE1YTNiNjdmP3M9NDgwJnI9cGcmZD1odHRwcyUzQSUyRiUyRmNkbi5hdXRoMC5jb20lMkZhdmF0YXJzJTJGYWMucG5nIiwidXBkYXRlZF9hdCI6IjIwMTgtMDQtMjZUMTY6MjA6MDguOTQ1WiIsImVtYWlsIjoiYWMtZGVtb0BhdXJlbGlhdG9vbHMuY29tIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJpc3MiOiJodHRwczovL2F1cmVsaWF0b29scy5hdXRoMC5jb20vIiwic3ViIjoiYXV0aDB8NWFlMWU5MTQ1ZmIzYmMxYjZkN2MyYzg4IiwiYXVkIjoiUldicWoydk9XNTRuQ2VUYzZKMW9PZ3dwaTdJNTh3N2oiLCJpYXQiOjE1MjQ3NTk2MDgsImV4cCI6MTUyNDc5NTYwOCwiYXRfaGFzaCI6IjYzTHFkV2pnQ1hDX1FDVnhOaHMxLVEiLCJub25jZSI6IjBnQ1NfLjRuM3FOb1RCeGZaT2V0TjZuTjdPVFlaVDBnIn0.QpOkBXtl3rc56dwRpeyvyglgqFrHsy9_nB-zTYZenkWV9qKOSgXl6hZCYu0uH2Vw6nGVaTx76zCQk2bNamnQIcLgHoCERLHxsLpJ1rqnJzu-CkCgj3SLLur6G2pXrjzIKVdBVA2a6hr4s1Ar3DNZLXtQWlypmC_tn2027oZOVkylyeupuRNgiJ98Tvfr0nY6wqlJI09x4AkDo8qvCsTUP250qrP2keiftwEerCBFD5FjP5yIrCEuAQcpyZS38Z-gBft9QWFk1v4gsOVQBG_gS13Cfox7nhvb9ehxlKPoCLaGd_CUWWj15hGPxTyOPenHDvgYIkxQky2q54kude_lIQ
```

**Comment by aaguiarz** [comment]
The application at http://localhost:3000/ should call parseHash on page load and that would get the fragment processed.

The error you are getting seems to indicate the app is not being hosted in http://localhost:3000

**Comment by adriatic** [comment]
Thanks for pointing out that I am feeble-minded, @aaguiarz - of course that the app is not hosted in http://localhost:3000 - the port should be 9000 as all aurelia sample applications that people use as a start for their own app, are. I simply saw 3000 and "parsed" if with my eyes as 9000.

Fixing this type, I went a lot further, what reminded me of the problems I encountered a year ago, when Auth0 code was in the transition to OpenID / Oauth 2.0 and I blamed the unstable Auth0 code for my lack of getting this done. 

So, here is my **actual** situation with the setup defined **[above](https://github.com/auth0-community/auth0-aurelia-samples/issues/22#issuecomment-384705580)**.

1. I cannot define the callback anywhere else except at root of my application, because the hash-fragment used by Auth0 to return the results of authentication process **interferes** with Aurelia's own use of hash tag in routing.

**Comment by adriatic** [comment]
My summary: before giving you the PRs for https://github.com/auth0-community/auth0-aurelia-samples I wanted to build the reasonably complex Aurelia application that uses Auth0-js API 😀 
- written in Typescript
- using SystemJS loader
- supporting the some basic state management that allows the authentication at some later time in app's life (say when the user selects a route that requires authentication and given access type).

**Comment by aaguiarz** [comment]
Hi @adriatic , where can I see the sample you are writing?

About the `req.setNotDelay is not a function` error, seems related to usage of JSPM (https://github.com/auth0/lock/issues/1167), is Aurelia using it?

**Comment by aaguiarz** [comment]
OK, about the `req.setNotDelay is not a function`, if you can send a small example that reproduces it would be helpful. It seems related to how `superagent` is loaded.
